soot-infoflow-android/src/soot/jimple/infoflow/android/iccta/IccInstrumenter.java

```diff
@@ -108,55 +114,101 @@ protected void instrumentMessenger() {
 			List<SootMethod> methodCopyList = new ArrayList<>(sootClass.getMethods());
 			for (SootMethod sootMethod : methodCopyList) {
 				if (sootMethod.isConcrete()) {
-					final Body body = sootMethod.retrieveActiveBody();
-					final LocalGenerator lg = Scene.v().createLocalGenerator(body);
+					final Body body = sootMethod.retrieveActiveBody();				
 
 					// Mark the method as processed
 					if (!processedMethods.add(sootMethod))
 						continue;
 
 					for (Iterator<Unit> unitIter = body.getUnits().snapshotIterator(); unitIter.hasNext();) {
-						Stmt stmt = (Stmt) unitIter.next();
-
-						if (stmt.containsInvokeExpr()) {
-							SootMethod callee = stmt.getInvokeExpr().getMethod();
-
-							// For Messenger.send(), we directly call the respective handler
-							if (callee == smMessengerSend) {
-								Set<SootClass> handlers = MessageHandler.v().getAllHandlers();
-								for (SootClass handler : handlers) {
-									Local handlerLocal = lg.generateLocal(handler.getType());
-
-									Unit newU = Jimple.v().newAssignStmt(handlerLocal,
-											Jimple.v().newNewExpr(handler.getType()));
-									newU.addTag(SimulatedCodeElementTag.TAG);
-									body.getUnits().insertAfter(newU, stmt);
-									instrumentedUnits.put(body, newU);
-
-									SootMethod initMethod = handler.getMethod("void <init>()");
-									Unit initU = Jimple.v().newInvokeStmt(
-											Jimple.v().newSpecialInvokeExpr(handlerLocal, initMethod.makeRef()));
-									initU.addTag(SimulatedCodeElementTag.TAG);
-									body.getUnits().insertAfter(initU, newU);
-									instrumentedUnits.put(body, initU);
+						Unit unit = unitIter.next();
+						Stmt stmt = (Stmt) unit;
+						//+++ collect handler fields signature	
+						if (stmt.containsFieldRef() && stmt.getFieldRef().getType().toString().contains("android.os.Handler")) {
```

Comment:
Please check whether the result of ``FieldRef.getType`` is a ``RefType``. On a ``RefType``, you can call ``getSootClass().getName()`` to get the proper name of the class. Don't expect ``toString()`` to be anything but debug information.